# Numpy

## 학습목표
- 데이터분석 라이브러리인 Numpy 에 대해서 알 수 있다.
- 수치계산에 특화된 Numpy 특성을 통해 데이터분석을 진행할 수 있다.
- list와의 차이점에 대해서 알 수 있다.

### Numpy
- Numerical Python
- 수치계산용 라이브러리
- 특징
  - 파이썬 List 자료형과 비슷한 형태 -> []
  - ndarray 클래스를 통해 생성 > array 라고 표현
  - ndarray(동일한 자료형) 배열형태로 존재
  - 각 값들은 순서를 가지는 데이터 > 정수로 index가 부여

In [2]:
import numpy as np

- 1차원 배열 (array) 생성

In [4]:
np.array([1,2,3]) # 여러개의 숫자는 list 형태로

array([1, 2, 3])

In [5]:
list1 = [1,2,3,4,5]

arr = np.array(list1)
arr

array([1, 2, 3, 4, 5])

- 2차원 배열 (array) 생성

In [6]:
list2 = [[1,2,3,4,5],[6,7,8,9,10]]

arr2 = np.array(list2)
arr2

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10]])

- numpy 배열에서 제공하는 다양한 기능 사용
  - 배열의 크기, 차원, 데이터타입 등 확인
  - *데이터를 분석하기 전에, 데이터가 형태 확인이 매우 중요함!*

- 정리
  - shape : 크기
  - ndim : 차원
  - dtype : 데이터타입(자료형)
  - size : 데이터요소개수

In [9]:
# 배열의 크기 확인 > shape 
arr2.shape
# (행,열)

(2, 5)

In [11]:
# 1차원의 크기 > 데이터 개수
arr.shape

(5,)

In [14]:
# 배열의 차원 확인 > ndim
print(arr2.ndim,arr.ndim)

2 1


In [17]:
# 배열의 데이터 타입 확인 > dtype (동일한 자료형을 사용하기 때문에)
print(arr.dtype,arr2.dtype)

int64 int64


In [19]:
# 데이터의 총 요소 개수 확인 > size
print(arr.size, arr2.size)

5 10


#### numpy 배열 연산

In [20]:
lst1 = [1,2,3]
lst2 = [4,5,6]

In [21]:
arr1 = np.array(lst1)
arr2 = np.array(lst2)

In [22]:
arr1,arr2

(array([1, 2, 3]), array([4, 5, 6]))

In [23]:
lst1 + lst2

[1, 2, 3, 4, 5, 6]

In [34]:
#list 연산을 통해 > list_result=[5,7,9]
# 만일 for문을 2번써서 계산했다면, 총 9번의 연산을 하게 되므로 잘못된 값이 나옴
list_result = [] # 밖에 선언해 줘야함! > for문 안에 들어가면 마지막 값만 들어가게 됨
for i in range(3):
    list_result.append(lst1[i]+lst2[i])
print(list_result)

[5, 7, 9]


In [33]:
arr1+arr2 

array([5, 7, 9])

- numpy 배열은 요소별 연산 가능 👍👍
  - index 번호가 같은 데이터끼리 연산

#### numpy 배열 indexing,slicing 
- 인덱싱 : 인덱스 번호를 활용해서 요소1개를 추출하는 방법
  - 변수명[인덱스번호]
- 슬라이싱 : 범위를 활용해 여러개의 요소를 추출하는 방법
  - 변수명[시작인덱스 : 끝인덱스+1]

In [35]:
# 인덱싱
arr = np.array([[1,2,3],[4,5,6]])
arr

array([[1, 2, 3],
       [4, 5, 6]])

In [39]:
print(arr[0][1])

2


(None, np.int64(2))

In [40]:
# [행,열]
arr[0,1]

np.int64(2)

In [42]:
# 슬라이싱
arr1 = np.array(range(10))
arr1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [43]:
# arange () : array + range > 일정한 규칙으로 증가하는 배열 생성
# arange(시작값, 끝값+1, step)
arr1 = np.arange(10)
arr1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [44]:
arr1[2:6]

array([2, 3, 4, 5])

In [45]:
arr1 + 3

array([ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [46]:
arr1 * 4

array([ 0,  4,  8, 12, 16, 20, 24, 28, 32, 36])

- array와 숫자로 연산을 하면 각 요소별로 연산

In [47]:
(arr1 % 2 == 0)

array([ True, False,  True, False,  True, False,  True, False,  True,
       False])

In [49]:
arr1[arr1 > 3]

array([4, 5, 6, 7, 8, 9])

- 반복문 최소화하여 연산을 빠르게 하기위해 > numpy library 사용

### 데이터 활용한 실습

- 예제 1

10명에 대한 키와 몸무게가 들어있는 파일을 읽어 각 사람별 BMI 지수를 구하시오

In [55]:
dt = np.loadtxt('data/height_weight.txt',delimiter=",") # delimiter > 구분자
dt

array([[175.2, 180.3, 175. , 169.2, 185.2, 188. , 177.6, 178.2, 177. ,
        179. ],
       [ 65.6,  88. ,  79.2,  69.3,  55. ,  71.2,  73. ,  68.9,  74. ,
         82. ]])

In [65]:
height_m = dt[0]/100

In [66]:
weight = dt[1]

In [67]:
BMI = weight/(height_m*height_m)
BMI

array([21.37153104, 27.07018468, 25.86122449, 24.20652885, 16.03543423,
       20.14486193, 23.14392095, 21.69720651, 23.62028791, 25.59220998])

In [69]:
weight/height_m**2

array([21.37153104, 27.07018468, 25.86122449, 24.20652885, 16.03543423,
       20.14486193, 23.14392095, 21.69720651, 23.62028791, 25.59220998])

- 색인 : 배열안의 원하는 데이터를 가져오기.
  - 인덱싱
  - 슬라이싱
  - boolean 색인 (요소의 개수 = size 가 동일해야함)
    - 원하는 조건에 일치하는 데이터만 확인

In [72]:
score = np.array([70,69,85,99,60])
b = [True,False,True,True,False]
score[b]

array([70, 85, 99])

In [74]:
b1 = score >= 80
b1

array([False, False,  True,  True, False])

In [135]:
score[:][b1]

array([85, 99])

In [75]:
score[b1]

array([85, 99])

In [76]:
score[score>=85]

array([85, 99])

### numpy 함수 
    - 합계 : 배열.sum() / np.sum(배열)
    - 평균 : 배열.mean() / np.mean(배열)
    - 절댓값 : np.abs(배열)
    - 가장 큰 값 : 배열.max() / np.max(배열)
    - 가장 작은 값 : 배열.min() / np.min(배열)
    - 중간값 : np.median(배열)
    - 실수값 반올림 : np.round(배열, 자릿수)

In [77]:
print(score.sum(), np.sum(score))

383 383


In [79]:
print(score.mean(), np.mean(score))

76.6 76.6


In [81]:
arr = np.array([-7,-5,1,4,6,-9])
np.abs(arr)

array([7, 5, 1, 4, 6, 9])

In [82]:
arr.argmax()

np.int64(4)

In [84]:
np.max(arr)

np.int64(6)

In [87]:
np.median(score)

np.float64(70.0)

In [88]:
arr2 = np.array([1.5502929,2.506098094,4.02493049])
np.round(arr2,1)

array([1.6, 2.5, 4. ])

#### 2. 영화 평점 데이터 분석하기

- 1 columns : user 번호
- 2 columns : 영화 id
- 3 columns : 영화평점
- 4 columns : 사용 XX

In [3]:
ratings = np.loadtxt('data/ratings.txt',delimiter=",",dtype='int')
ratings

array([[        1,      1193,         5, 978300760],
       [        1,       661,         3, 978302109],
       [        1,       914,         3, 978301968],
       ...,
       [     6040,       562,         5, 956704746],
       [     6040,      1096,         4, 956715648],
       [     6040,      1097,         4, 956715569]], shape=(1000209, 4))

In [103]:
print(ratings.ndim,ratings.shape,ratings.size)

2 (1000209, 4) 4000836


- 2.1 전체 영화의 평점 평균 구하기

In [119]:
print(ratings[:,2].mean(),ratings[:,2].mean().round(2))

3.581564453029317 3.58


In [118]:
print(np.mean(ratings[:,2]))

3.581564453029317


- 2.2 사용자 아이디가 1인 평점 평균 구하기

In [124]:
print(ratings[ratings[:,0] == 1,2].mean().round(2))

4.19


In [152]:
ratings[ratings[:,0] == 1][:,2].mean().round(2)

np.float64(4.19)

In [150]:
ratings[:,2][ratings[:,0]==2].mean().round(2)

np.float64(3.71)

In [ ]:
for i in range(

- 2.3 모든 사용자의 평점 평균
  - 중간에 누락된 사람이 있을 수도 있음.. > 그냥 무조건 for문을 돌리면 안됨!
  - 중복된 값들 중에서 unique한 값들만 뽑아서 확인한 후, 전체 계산 >


  ```
  np.unique(배열)
  ```

- 처음 그냥 짰던 코드

In [137]:
l = []
for i in range(1,ratings[:,0].max()+1):
    l.append(ratings[ratings[:,0] == i,2].mean().round(2))

In [189]:
l[:5]

[np.float64(4.19),
 np.float64(3.71),
 np.float64(3.9),
 np.float64(4.19),
 np.float64(3.15)]

In [23]:
unique_user_id = np.unique(ratings[:,0]) # 사람은 누락된 번호가 없음
# for userId in unique_user_id:
#     print(userId)

In [24]:
unique_movie_id = np.unique(ratings[:,1]) # 영화 ID는 누락된 번호가 존재함!

In [6]:
# unique한 user

In [7]:
int(unique_user_id[0])

1

In [8]:
person_ratings = []
for i in range(1,len(unique_user_id)+1):
    person_ratings.append(ratings[ratings[:,0]==i,2].mean())
    

- 밑에 처럼 코드 짜면...6040*6040 만큼 반복이 돌아감...

```
result = []
person_ratings = []

for i in range(1,len(unique_user_id)+1):
    for i in range(1,len(unique_user_id)+1):
        person_ratings.append(int(unique_user_id[i-1]))
        person_ratings.append(float(ratings[ratings[:,0]==i,2].mean()))
    result.append(person_ratings)
```

- 각각 사람별로 list가 만들어지지 않음...

In [204]:
result = []
person_ratings = []
for i in range(1,len(unique_user_id)+1):
    person_ratings.append(int(unique_user_id[i-1]))
    person_ratings.append(float(ratings[ratings[:,0]==i,2].mean()))
    

- 그래서 각 변수로 받아서, 리스트 자체를 append > append안에도 list를 넣을 수 있다!

In [30]:
# result = []
person_ratings = []
for i in range(1,len(unique_user_id)+1):
    person = (int(unique_user_id[i-1]))
    person_mean = (float(ratings[ratings[:,0]==i,2].mean()))
    person_ratings.append([person,person_mean]) # append 안에도 list자체를 그냥 넣으면 됨!

In [31]:
person_ratings[:5]
    

[[1, 4.188679245283019],
 [2, 3.7131782945736433],
 [3, 3.9019607843137254],
 [4, 4.190476190476191],
 [5, 3.1464646464646466]]

- for문을 range로 하지 않고, 리스트 자체로 받아오면 해결하기가 용이.. (다시 코드 처음부터 생각해보기!!!)

In [19]:
userList = []
for userId in unique_user_id:
    user = float(ratings[ratings[:,0] == userId,2].mean())
    userList.append([int(userId),user])

In [20]:
userList[:5]

[[1, 4.188679245283019],
 [2, 3.7131782945736433],
 [3, 3.9019607843137254],
 [4, 4.190476190476191],
 [5, 3.1464646464646466]]

- 2.4 영화별 전체 평점 구하기

In [26]:
unique_movie_id

array([   1,    2,    3, ..., 3950, 3951, 3952], shape=(3706,))

In [42]:
movie_mean_ = (float(ratings[ratings[:,1]==1,2].mean()))

In [43]:
movie_mean_

4.146846413095811

- 이렇게 계산하면 똑같은 평균값만 들어감 > 에러난 이유.. ```ratings[ratings[:,1]==i``` 가 아니다...

In [37]:
movie_List = []
for movieId in unique_movie_id:
    if movieId in unique_movie_id:
        movie_mean = (float(ratings[ratings[:,1]==i,2].mean())) #i가 아님!!!!
        movie_List.append([int(movieId),movie_mean])
# 

In [38]:
movie_List

[[1, 3.404017857142857],
 [2, 3.404017857142857],
 [3, 3.404017857142857],
 [4, 3.404017857142857],
 [5, 3.404017857142857],
 [6, 3.404017857142857],
 [7, 3.404017857142857],
 [8, 3.404017857142857],
 [9, 3.404017857142857],
 [10, 3.404017857142857],
 [11, 3.404017857142857],
 [12, 3.404017857142857],
 [13, 3.404017857142857],
 [14, 3.404017857142857],
 [15, 3.404017857142857],
 [16, 3.404017857142857],
 [17, 3.404017857142857],
 [18, 3.404017857142857],
 [19, 3.404017857142857],
 [20, 3.404017857142857],
 [21, 3.404017857142857],
 [22, 3.404017857142857],
 [23, 3.404017857142857],
 [24, 3.404017857142857],
 [25, 3.404017857142857],
 [26, 3.404017857142857],
 [27, 3.404017857142857],
 [28, 3.404017857142857],
 [29, 3.404017857142857],
 [30, 3.404017857142857],
 [31, 3.404017857142857],
 [32, 3.404017857142857],
 [33, 3.404017857142857],
 [34, 3.404017857142857],
 [35, 3.404017857142857],
 [36, 3.404017857142857],
 [37, 3.404017857142857],
 [38, 3.404017857142857],
 [39, 3.4040178571428

In [46]:
# 완
movie_List = []
for movieId in unique_movie_id:
    if movieId in unique_movie_id:
        movie_mean = (float(ratings[ratings[:,1]==movieId,2].mean())) #i가 아님!!!!
        movie_List.append([int(movieId),movie_mean])


In [48]:
movie_List[:5]


[[1, 4.146846413095811],
 [2, 3.20114122681883],
 [3, 3.01673640167364],
 [4, 2.7294117647058824],
 [5, 3.0067567567567566]]

- movie Id가 unique_movie_id에 있을 때만, 평균을 계산해서 

In [53]:
movie_List_ = []

if movieId in unique_movie_id:
    for movieId in unique_movie_id:
        movie_mean = (float(ratings[ratings[:,1]==movieId,2].mean()))
        movie_List_.append([int(movieId),movie_mean])

In [54]:
movie_List_[:5]

[[1, 4.146846413095811],
 [2, 3.20114122681883],
 [3, 3.01673640167364],
 [4, 2.7294117647058824],
 [5, 3.0067567567567566]]

In [29]:
float(ratings[ratings[:,1] == 1,2].mean())

4.146846413095811

In [32]:
# 예외처리 안하고 코딩
movie_ratings = []
for i in range(1,len(unique_movie_id)+1):
    movie = (int(unique_movie_id[i-1]))
    movie_mean = (float(ratings[ratings[:,1]==i,2].mean()))
    movie_ratings.append([movie,movie_mean]) # append 안에도 list자체를 그냥 넣으면 됨!

C:\Users\ChoiSeokjin\AppData\Local\Temp\ipykernel_11932\86065245.py:5: RuntimeWarning: Mean of empty slice.
  movie_mean = (float(ratings[ratings[:,1]==i,2].mean()))
C:\Users\ChoiSeokjin\anaconda3\Lib\site-packages\numpy\_core\_methods.py:144: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)


In [33]:
movie_ratings[:5]

[[1, 4.146846413095811],
 [2, 3.20114122681883],
 [3, 3.01673640167364],
 [4, 2.7294117647058824],
 [5, 3.0067567567567566]]

In [55]:
# 예외처리 작성해서 
movie_ratings_2 = []
for i in range(1,len(unique_movie_id)+1):
    if ratings[i,1] in unique_movie_id:
        movie = (int(unique_movie_id[i-1]))
        movie_mean = (float(ratings[ratings[:,1]==i,2].mean()))
        movie_ratings_2.append([movie,movie_mean]) # append 안에도 list자체를 그냥 넣으면 됨!

C:\Users\ChoiSeokjin\AppData\Local\Temp\ipykernel_11932\768684543.py:6: RuntimeWarning: Mean of empty slice.
  movie_mean = (float(ratings[ratings[:,1]==i,2].mean()))


In [56]:
movie_ratings_2[:5]

[[1, 4.146846413095811],
 [2, 3.20114122681883],
 [3, 3.01673640167364],
 [4, 2.7294117647058824],
 [5, 3.0067567567567566]]

- i 값 자체가 unique_movie_id 인덱스를 의미하므로 i를 바로 비교해도 괜찮다! > 에러도 안남!

In [60]:
# 예외처리 작성해서 
movie_ratings_2 = []
for i in range(1,len(unique_movie_id)+1):
    if i in unique_movie_id:
        movie = (int(unique_movie_id[i-1]))
        movie_mean = (float(ratings[ratings[:,1]==i,2].mean()))
        movie_ratings_2.append([movie,movie_mean]) # append 안에도 list자체를 그냥 넣으면 됨!

In [61]:
movie_ratings_2[:5]

[[1, 4.146846413095811],
 [2, 3.20114122681883],
 [3, 3.01673640167364],
 [4, 2.7294117647058824],
 [5, 3.0067567567567566]]

In [160]:
# 색인 연습 > 행 먼저 , 열은 다음에
arr3 = np.arange(50).reshape(5,10)
arr3

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49]])

In [163]:
arr3[:2,:]

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19]])

In [162]:
arr3[:2]

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19]])

In [164]:
arr3[:2,2:6]

array([[ 2,  3,  4,  5],
       [12, 13, 14, 15]])

In [167]:
arr3[:,7]

(5,)

In [168]:
arr3[2,2:8]

array([22, 23, 24, 25, 26, 27])

In [172]:
arr3[arr3[:,0]>=20]

array([[20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49]])

In [173]:
arr3[arr3[:,0]>=20,5:7]

array([[25, 26],
       [35, 36],
       [45, 46]])